src/ethereum/frontier/vm/instructions/environment.py

```diff
@@ -0,0 +1,244 @@
+"""
+Ethereum Virtual Machine (EVM) Environmental Instructions
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. contents:: Table of Contents
+    :backlinks: none
+    :local:
+
+Introduction
+------------
+
+Implementations of the EVM environment related instructions.
+"""
+
+from ethereum.base_types import U256, Uint
+
+from .. import Evm
+from ..gas import GAS_BASE, GAS_EXT, GAS_VERY_LOW, subtract_gas
+from ..stack import pop, push
+
+# Address can be only 20 bytes
+MAX_ADDRESS_BYTES = 20
+
+
+def address(evm: Evm) -> None:
+    """
+    Pushes the address of the current executing account to the stack.
+
+    Parameters
+    ----------
+    evm :
+        The current EVM frame.
+
+    Raises
+    ------
+    OutOfGasError
+        If `evm.gas_left` is less than `2`.
+    """
+    evm.gas_left = subtract_gas(evm.gas_left, GAS_BASE)
+    push(evm.stack, U256.from_be_bytes(evm.current))
+
+
+def balance(evm: Evm) -> None:
+    """
+    Pushes the balance of the given account onto the stack.
+
+    Parameters
+    ----------
+    evm :
+        The current EVM frame.
+
+    Raises
+    ------
+    StackUnderflowError
+        If `len(stack)` is less than `1`.
+    OutOfGasError
+        If `evm.gas_left` is less than `20`.
+    """
+    # TODO: There are no test cases against this function. Need to write
+    # custom test cases.
+    evm.gas_left = subtract_gas(evm.gas_left, GAS_EXT)
+
+    address = pop(evm.stack).to_be_bytes()
+    if len(address) > MAX_ADDRESS_BYTES:
+        # Get the last 20 bytes and consider this as the address
+        address = address[-MAX_ADDRESS_BYTES:]
+
+    account_state = evm.env.state.get(address, None)
```

Comment:
Hmmmm... Should `state` be a `defaultdict` instead of a `dict`?

Reply:
I guess a better way to phrase that question would be: are there cases where `None` and a default `Account` need to be distinguished, or should they be indistinguishable?

Reply:
I feel using defaultdict can sometimes get dangerous. For example it's common to do things like follow which will grow the size of the state unknowingly and unintentionally.
```
In [1]: from collections import defaultdict

In [2]: ob = defaultdict(int)

In [3]: ob["key1"]
Out[3]: 0

In [4]: ob
Out[4]: defaultdict(int, {'key1': 0})

In [5]: try:
   ...:     ob["key2"]
   ...: except KeyError:
   ...:     pass
   ...: 

In [6]: ob
Out[6]: defaultdict(int, {'key1': 0, 'key2': 0})    # Unintentionally increased the object size.

```

That said, I'm not sure if it is worth it and if it adds to the code complexity (readability wise). Let me know your thoughts. Thanks.

Reply:
This doesn't need to be decided now, I was mostly musing about whether this is a good fit for `defaultdict`. Good point about unintentionally increasing the size.